File: 3525.find-x-value-of-array-ii.py
#
# @lc app=leetcode id=3525 lang=python3
#
# [3525] Find X Value of Array II
#
# @lc code=start
class Solution:
    def resultArray(self, nums: List[int], k: int, queries: List[List[int]]) -> List[int]:
        result = []
        n = len(nums)
        
        for query in queries:
            indexi, valuei, starti, xi = query
            # Update nums with new value at indexi
            nums[indexi] = valuei
            
            # Calculate x-value for current configuration of nums[starti:]
            current_x_value = self.calculate_x_value(nums[starti:], k, xi)
            result.append(current_x_value)
        return result
    
    def calculate_x_value(self, nums: List[int], k: int, x: int) -> int:
        count_ways = 0
        current_product_mod_k = 1  # Product modulo k of current suffix elements
        
        # Calculate number of valid suffix removals by iterating from end to start
        for num in reversed(nums):
            if current_product_mod_k == x:
                count_ways += 1
            current_product_mod_k = (current_product_mod_k * num) % k
        if current_product_mod_k == x:  # Check full product case as well
            count_ways += 1
        return count_ways
# @lc code=end

File: 3525.find-x-value-of-array-ii.go
#
# @lc app=leetcode id=3525 lang=golang
#
# [3525] Find X Value of Array II
#

# @lc code=start
func resultArray(nums []int, k int, queries [][]int) []int {
    n := len(nums)
    res := make([]int, len(queries))
    curr := make([]int, n)
    copy(curr, nums)
    for qi, q := range queries {
        idx, val, start, x := q[0], q[1], q[2], q[3]
        curr[idx] = val
        arr := curr[start:]
        m := len(arr)
        prefixProd := make([]int, m+1)
        prefixProd[0] = 1
        for i := 0; i < m; i++ {
            prefixProd[i+1] = (prefixProd[i] * arr[i]) % k
        }
        count := 0
        for j := 1; j <= m; j++ {
            if prefixProd[j] == x {
                count++
            }
        }
        res[qi] = count
    }
    return res
}
# @lc code=end